src/main/java/com/mongodb/jdbc/utils/X509Authentication.java

```diff
@@ -0,0 +1,144 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.jdbc.utils;
+
+import com.mongodb.jdbc.logging.MongoLogger;
+import java.io.FileReader;
+import java.security.*;
+import java.security.cert.Certificate;
+import java.util.logging.Level;
+import javax.net.ssl.KeyManagerFactory;
+import javax.net.ssl.SSLContext;
+import org.bouncycastle.asn1.pkcs.PrivateKeyInfo;
+import org.bouncycastle.cert.X509CertificateHolder;
+import org.bouncycastle.cert.jcajce.JcaX509CertificateConverter;
+import org.bouncycastle.jce.provider.BouncyCastleProvider;
+import org.bouncycastle.openssl.PEMParser;
+import org.bouncycastle.openssl.jcajce.JcaPEMKeyConverter;
+import org.bouncycastle.pkcs.PKCS8EncryptedPrivateKeyInfo;
+import org.bouncycastle.pkcs.jcajce.JcePKCSPBEInputDecryptorProviderBuilder;
+
+public class X509Authentication {
+    private static final String BOUNCY_CASTLE = "BC";
+    private final MongoLogger logger;
+
+    static {
+        if (Security.getProvider(BOUNCY_CASTLE) == null) {
+            Security.addProvider(new BouncyCastleProvider());
+        }
+    }
+
+    public X509Authentication(MongoLogger logger) {
+        this.logger = logger;
+    }
+
+    public void configureX509Authentication(
+            com.mongodb.MongoClientSettings.Builder settingsBuilder,
+            String pemPath,
+            char[] passphrase) {
+
+        logger.log(Level.FINE, "Using client certificate for X509 authentication: " + pemPath);
+        if (passphrase != null && passphrase.length > 0) {
+            logger.log(Level.FINE, "Client certificate passphrase has been specified");
+        }
+        try {
+            SSLContext sslContext = createSSLContext(pemPath, passphrase);
+
+            settingsBuilder.applyToSslSettings(
+                    sslSettings -> {
+                        sslSettings.enabled(true);
+                        sslSettings.context(sslContext);
+                    });
+        } catch (Exception e) {
+            throw new RuntimeException("SSL setup failed", e);
+        }
+    }
+
+    private SSLContext createSSLContext(String pemPath, char[] passphrase) throws Exception {
+        PrivateKey privateKey = null;
+        Certificate cert = null;
+
+        try (PEMParser pemParser = new PEMParser(new FileReader(pemPath))) {
+            Object pemObj;
+
+            // Iterate through PEM objects found in the PEM file and process them based on type:
+            //  - Encrypted/unencrypted private keys
+            //  - X.509 certificates
+            while ((pemObj = pemParser.readObject()) != null) {
+                try {
+                    if (passphrase != null
+                            && passphrase.length > 0
+                            && pemObj instanceof PKCS8EncryptedPrivateKeyInfo) {
```

Comment:
Does this mean that only PKCS8 is supported? Not PKCS12?

Reply:
Yes, the PEM file format doesn't support PKCS12 I would need to add logic to support reading the PKCS12 file since that's a binary `*.p12` or `*.pks` file.  I focused more on PEM since that's what we get from the Atlas UI and looking up the self-signed [documentation](https://www.mongodb.com/docs/manual/tutorial/configure-x509-client-authentication/) I see PEM files used.  
I can create a ticket to support pkcs12 if that's what we want.  But even with regard to PEM, that's me speculating on what the customer will want to use.

Reply:
Sounds good to me. Especially since this is what users can download on Atlas.
We might need to make the doc very clear about this, that's all 👍